internal/webhook/multiclusterservice_webhook.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2024
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webhook
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/Mirantis/hmc/api/v1alpha1"
+	"github.com/Mirantis/hmc/internal/utils"
+)
+
+type MultiClusterServiceValidator struct {
+	client.Client
+}
+
+const invalidMultiClusterServiceMsg = "the MultiClusterService is invalid"
+
+// SetupWebhookWithManager will setup the manager to manage the webhooks
+func (v *MultiClusterServiceValidator) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	v.Client = mgr.GetClient()
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&v1alpha1.MultiClusterService{}).
+		WithValidator(v).
+		WithDefaulter(v).
+		Complete()
+}
+
+var (
+	_ webhook.CustomValidator = &MultiClusterServiceValidator{}
+	_ webhook.CustomDefaulter = &MultiClusterServiceValidator{}
+)
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type.
+func (*MultiClusterServiceValidator) Default(_ context.Context, _ runtime.Object) error {
+	return nil
+}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (v *MultiClusterServiceValidator) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	mcs, ok := obj.(*v1alpha1.MultiClusterService)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected MultiClusterService but got a %T", obj))
+	}
+
+	if err := validateServices(ctx, v.Client, utils.DefaultSystemNamespace, mcs.Spec.Services); err != nil {
+		return nil, fmt.Errorf("%s: %w", invalidMultiClusterServiceMsg, err)
+	}
+
+	return nil, nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (v *MultiClusterServiceValidator) ValidateUpdate(ctx context.Context, _, newObj runtime.Object) (admission.Warnings, error) {
+	mcs, ok := newObj.(*v1alpha1.MultiClusterService)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected MultiClusterService but got a %T", newObj))
+	}
+
+	if err := validateServices(ctx, v.Client, utils.DefaultSystemNamespace, mcs.Spec.Services); err != nil {
```

Comment:
The `utils.DefaultSystemNamespace` is the default namespace where HMC can be installed if not overwritten (see https://github.com/eromanova/hmc/blob/main/internal/utils/kube.go#L52. The actual system namespace may be different. You should pass the real current namespace to the validator and use it here instead.

Reply:
The same for https://github.com/eromanova/hmc/blob/main/internal/controller/multiclusterservice_controller.go#L68. It'll not work if HMC is installed into another namespace. But it can be fixed in a separate PR.